presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.sql.analyzer.Analysis;
+import com.facebook.presto.sql.analyzer.TypeSignatureProvider;
+import com.facebook.presto.sql.tree.ArrayConstructor;
+import com.facebook.presto.sql.tree.Cast;
+import com.facebook.presto.sql.tree.Cube;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.GroupingElement;
+import com.facebook.presto.sql.tree.GroupingOperation;
+import com.facebook.presto.sql.tree.GroupingSets;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.facebook.presto.sql.tree.Rollup;
+import com.facebook.presto.type.ListLiteralType;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.operator.scalar.GroupingOperationFunction.GROUPING;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.resolveFunction;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class GroupingOperationRewriter
```

Comment:
I don't think applying `GroupingOperationRewriter` in a lot of places (and pushing required arguments) is maintainable in a long run. I think the approach that is the most compatible with analyzer/planner is as follows:
1. Add `IdentityHashMap<QuerySpecifictaion, List<GroupingOperation>> groupingOperations` to `Analysis`.
2. Add `analyzeGroupings` method to `StatementAnalyzer`.
   * executed after/before `analyzeAggregations` during `visitQuerySpecification`
   * performs grouping expressions validation (see also: https://github.com/prestodb/presto/pull/6704/commits/38db2c8e0f9b04a4b25f8876cc9fecd5af5a6656#r103871453)
   * puts extracted groupings expressions for `QuerySpecification` to `Analysis`
3. Add `PlanBuilder handleGroupings(PlanBuilder subPlan, QuerySpecification node, Iterable<GroupingOperation> groupings)` method to `QueryPlanner`
   - called after: https://github.com/prestodb/presto/pull/6704/commits/38db2c8e0f9b04a4b25f8876cc9fecd5af5a6656#diff-21d31cb9d487772413eff4f67b8555deR154
   - it should add projection that will push-though existing symbols and also compute `grouping(..)` expressions
     * you can probably use `PlanBuilder#appendProjections`
   - it should add mappings from `groupings(...)` AST expressions to symbols in `TranslationMap`, e.g:
     * `translationMap.put(groupingExpression, symbol)`

With that approach using rewriter in multiple places is no longer. `TranslationMap` will correctly rewrite `grouping(...)` AST expressions to symbols